File: apps/api/app/routers/builds.py
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel

from app.auth import get_current_user_phone
from app.deps import optional_supabase

router = APIRouter(prefix="/builds", tags=["builds"])


class BuildRequest(BaseModel):
    prompt: str
    phone_number: str | None = None


@router.post("")
async def trigger_build(
    body: BuildRequest,
    phone: str = Depends(get_current_user_phone),
    db=Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    row = db.table("linq_jobs").insert({
        "user_phone": phone,
        "prompt": body.prompt,
        "status": "queued",
    }).execute()
    data = row.data[0] if row.data else {}
    return {"build_id": data.get("id"), "status": data.get("status")}


@router.get("/{build_id}")
async def get_build(
    build_id: str,
    phone: str = Depends(get_current_user_phone),
    db=Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    row = (
        db.table("linq_jobs")
        .select("*")
        .eq("id", build_id)
        .eq("user_phone", phone)
        .maybe_single()
        .execute()
    )
    if not row.data:
        raise HTTPException(404, "Build not found")
    return row.data


@router.get("")
async def list_builds(
    phone: str = Depends(get_current_user_phone),
    db=Depends(optional_supabase),
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    row = (
        db.table("linq_jobs")
        .select("*")
        .eq("user_phone", phone)
        .order("created_at", desc=True)
        .range(offset, offset + limit - 1)
        .execute()
    )
    return {"data": row.data, "limit": limit, "offset": offset}

File: apps/worker/src/routes/builds.ts
import { Hono } from 'hono'
import type { Env } from '../index'
import { getPhone } from '../middleware/auth'
import { getSupabase } from '../lib/supabase'

const app = new Hono<{ Bindings: Env }>()

app.post('/', async (c) => {
  let phone: string
  try { phone = await getPhone(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const { prompt, phone_number } = await c.req.json<{ prompt: string; phone_number?: string }>()
  const sb = getSupabase(c.env)
  const { data, error } = await sb.from('linq_jobs').insert({
    user_phone: phone,
    prompt,
    ...(phone_number ? { phone_number } : {}),
    status: 'queued',
  }).select().single()
  if (error) return c.json({ error: error.message }, 503)
  return c.json({ build_id: data.id, status: data.status })
})

app.get('/', async (c) => {
  let phone: string
  try { phone = await getPhone(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const limit = Math.min(parseInt(c.req.query('limit') ?? '20'), 100)
  const offset = parseInt(c.req.query('offset') ?? '0')
  const sb = getSupabase(c.env)
  const { data, error } = await sb
    .from('linq_jobs')
    .select('*')
    .eq('user_phone', phone)
    .order('created_at', { ascending: false })
    .range(offset, offset + limit - 1)
  if (error) return c.json({ error: error.message }, 503)
  return c.json(data)
})

app.get('/:id', async (c) => {
  let phone: string
  try { phone = await getPhone(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const sb = getSupabase(c.env)
  const { data, error } = await sb
    .from('linq_jobs')
    .select('*')
    .eq('id', c.req.param('id'))
    .eq('user_phone', phone)
    .maybeSingle()
  if (error) return c.json({ error: error.message }, 503)
  if (!data) return c.json({ error: 'Build not found' }, 404)
  return c.json(data)
})

export default app
